File: Java/src/_0052_N_Queens_II/Solution.java
package _0052_N_Queens_II;

import java.util.List;

/**
 * Date: 2020/9/3 15:15
 * Content:
 */
public class Solution {
    int totalNum;
    boolean[] col;
    boolean[] dia1;
    boolean[] dia2;

    private void totalNQueens(int n, int index) {
        if (index == n) {
            totalNum++;
            return;
        }

        for (int j = 0; j < n; j++)
            if (!col[j] && !dia1[index + j] && !dia2[index - j + n - 1]) {
                col[j] = true;
                dia1[index + j] = true;
                dia2[index - j + n - 1] = true;
                totalNQueens(n, index + 1);
                col[j] = false;
                dia1[index + j] = false;
                dia2[index - j + n - 1] = false;
            }
    }

    public int totalNQueens(int n) {
        col = new boolean[n];
        dia1 = new boolean[2 * n - 1];
        dia2 = new boolean[2 * n - 1];

        totalNQueens(n, 0);

        return totalNum;
    }

    public static void main(String[] args) {
        int n = 4;
        System.out.println((new Solution()).totalNQueens(n));
    }
}

File: Python/_0052_N_Queens_II/Solution.py
# coding=utf-8
# Date: 2021/5/24 10:33
from typing import List


class Solution:
    def totalNQueens(self, n: int) -> int:
        col = [False for _ in range(n)]
        dg = [False for _ in range(2 * n - 1)]
        udg = [False for _ in range(2 * n - 1)]

        return self.dfs(0, n, col, dg, udg)

    def dfs(self, u, n, col, dg, udg):
        if u == n:
            return 1
        res = 0
        for i in range(n):
            if not col[i] and not dg[u - i + n - 1] and not udg[u + i]:
                col[i] = dg[u - i + n - 1] = udg[u + i] = True
                res += self.dfs(u + 1, n, col, dg, udg)
                col[i] = dg[u - i + n - 1] = udg[u + i] = False
        return res


if __name__ == "__main__":
    print(Solution().totalNQueens(4))  # 2
